File: src/main/java/discretemath/combination/BitCombinations.java
package discretemath.combination;

import discretemath.bool.BooleanTable;
import discretemath.common.Direction;

public class BitCombinations {

	public static BooleanTable nBitCombinations(int n) {
		int rowsCount = 1 << n;

		BooleanTable table = BooleanTable.create(rowsCount, n);

		nBitCombinations(n, table, 0, 0, Direction.FROM_UP_TO_DOWN);

		return table;
	}

	public static void nBitCombinations(int n, BooleanTable table) {
		nBitCombinations(n, table, 0, 0, Direction.FROM_UP_TO_DOWN);
	}

	public static void nBitCombinations(int n, BooleanTable table, int startRow, int startColumn,
										Direction tableFillDirection) {
		TableCombinationValueSetter valueSetter = getTableValueSetter(tableFillDirection, startRow, startColumn);

		int rowsCount = 1 << n;

		for (int i = 0; i < rowsCount; i++) {
			for (int j = 0; j < n; j++) {
				boolean value = ((i >> j) & 1) == 1;
				valueSetter.perform(table, i, n - 1 - j, value);
			}
		}
	}

	private static TableCombinationValueSetter getTableValueSetter(Direction direction, int startRow, int startColumn) {
		return switch (direction) {
			case FROM_UP_TO_DOWN -> (table, i, j, value) -> table.set(startRow + i, startColumn + j, value);
			case FROM_LEFT_TO_RIGHT -> (table, i, j, value) -> table.set(startRow + j, startColumn + i, value);
		};
	}

	private interface TableCombinationValueSetter {

		void perform(BooleanTable table, int i, int j, boolean value);
	}
}

File: src/main/scala/discretemath/combination/BitCombinations.scala
package discretemath.combination

import discretemath.common.Direction
import discretemath.common.Direction.FROM_UP_TO_DOWN
import discretemath.structure.matrix.{ArrayMatrix, Matrix}

object BitCombinations {

  def nBitCombinations(n: Int): Matrix[Boolean] = {
    val rowsCount = 1 << n
    val table: Matrix[Boolean] = ArrayMatrix.create(rowsCount, n)
    nBitCombinations(n, table, 0, 0, Direction.FROM_UP_TO_DOWN, identity)
    table
  }

  def nBitCombinations(n: Int, table: Matrix[Boolean]): Unit = nBitCombinations(n, table, transformer = identity)

  def nBitCombinations[T](n: Int, table: Matrix[T], startRow: Int = 0, startColumn: Int = 0, tableFillDirection: Direction = FROM_UP_TO_DOWN, transformer: Boolean => T): Unit = {
    val valueSetter = getTableValueSetter(tableFillDirection, startRow, startColumn, transformer)
    val rowsCount = 1 << n
    for (i <- 0 until rowsCount) {
      for (j <- 0 until n) {
        val value = ((i >> j) & 1) == 1
        valueSetter(table, i, n - 1 - j, value)
      }
    }
  }

  private def getTableValueSetter[T](direction: Direction, startRow: Int, startColumn: Int, transformer: Boolean => T) = direction match {
    case Direction.FROM_UP_TO_DOWN => (table: Matrix[T], i: Int, j: Int, value: Boolean) => table.set(startRow + i, startColumn + j, transformer(value))
    case Direction.FROM_LEFT_TO_RIGHT => (table: Matrix[T], i: Int, j: Int, value: Boolean) => table.set(startRow + j, startColumn + i, transformer(value))
  }
}